source/vstd/raw_ptr.rs

```diff
@@ -0,0 +1,296 @@
+//
+// TODO flesh out the metadata system for working with DSTs
+// It may make sense to use <T as Pointee>::Metadata directly.
+#[verifier::external_body]
+pub ghost struct Metadata {}
+
+impl Metadata {
+    // Unit metadata used for thin pointers
+    pub spec fn unit() -> Metadata;
+}
+
+pub ghost struct PtrData {
+    pub addr: usize,
+    pub provenance: Provenance,
+    pub metadata: Metadata,
+}
+
+#[verifier(external_body)]
+#[verifier::accept_recursive_types(T)]
+pub tracked struct PointsTo<T> {
+    phantom: core::marker::PhantomData<T>,
+    no_copy: NoCopy,
+}
+
+// Don't use std Option here in order to avoid circular dependency issues
+// with verifying the standard library.
+// (Also, using our own enum here lets us have more meaningful
+// variant names like Uninit/Init.)
+#[verifier::accept_recursive_types(T)]
+pub ghost enum MemContents<T> {
+    Uninit,
+    Init(T),
+}
+
+pub ghost struct PointsToData<T> {
+    pub ptr: *mut T,
+    pub opt_value: MemContents<T>,
+}
+
+impl<T> View for *mut T {
+    type V = PtrData;
+
+    spec fn view(&self) -> Self::V;
+}
+
+impl<T> View for *const T {
+    type V = PtrData;
+
+    #[verifier::inline]
+    open spec fn view(&self) -> Self::V {
+        (*self as *mut T).view()
+    }
+}
+
+impl<T> View for PointsTo<T> {
+    type V = PointsToData<T>;
+
+    spec fn view(&self) -> Self::V;
+}
+
+impl<T> PointsTo<T> {
+    #[verifier::inline]
+    pub open spec fn ptr(&self) -> *mut T {
+        self.view().ptr
+    }
+
+    #[verifier::inline]
+    pub open spec fn opt_value(&self) -> MemContents<T> {
+        self.view().opt_value
+    }
+
+    #[verifier::inline]
+    pub open spec fn is_init(&self) -> bool {
+        self.opt_value().is_init()
+    }
+
+    #[verifier::inline]
+    pub open spec fn is_uninit(&self) -> bool {
+        self.opt_value().is_uninit()
+    }
+
+    #[verifier::inline]
+    pub open spec fn value(&self) -> T {
+        self.opt_value().value()
+    }
+}
+
+impl<T> MemContents<T> {
+    #[verifier::inline]
+    pub open spec fn is_init(&self) -> bool {
+        self is Init
+    }
+
+    #[verifier::inline]
+    pub open spec fn is_uninit(&self) -> bool {
+        self is Uninit
+    }
+
+    #[verifier::inline]
+    pub open spec fn value(&self) -> T {
+        self->0
+    }
+}
+
+//////////////////////////////////////
+// Inverse functions:
+// Pointers are equivalent to their model
+pub spec fn ptr_mut_from_data<T: ?Sized>(data: PtrData) -> *mut T;
+
+#[verifier::inline]
+pub open spec fn ptr_from_data<T: ?Sized>(data: PtrData) -> *const T {
+    ptr_mut_from_data(data) as *const T
+}
+
+#[verifier::external_body]
+pub broadcast proof fn axiom_ptr_mut_from_data<T>(data: PtrData)
+    ensures
+        (#[trigger] ptr_mut_from_data::<T>(data))@ == data,
+{
+}
+
+#[verifier::external_body]
+pub broadcast proof fn ptrs_mut_eq<T>(a: *mut T, b: *mut T)
+    requires
+        #[trigger] a@ == #[trigger] b@,
```

Comment:
This leads to a quadratic number of quantifier instantiations.  There's a standard trick to reduce this to a linear number of instantiations based on SMT uninterpreted function equality:

```
pub spec fn reverse_view_mut<T>(p: PtrData) -> *mut T;
pub spec fn reverse_view_const<T>(p: PtrData) -> *const T;

#[verifier::external_body]
pub broadcast proof fn ptrs_mut_eq<T>(a: *mut T)
    ensures
        reverse_view_mut::<T>(#[trigger] (a@)) == a,
{
}

#[verifier::external_body]
pub broadcast proof fn ptrs_eq<T>(a: *const T)
    ensures
        reverse_view_const::<T>(#[trigger] (a@)) == a,
{
}
```